Small fixes: use /dev instead of /tmp for mailbox, fixed unmapmem() to avoid exit(-1) at ws2811_fini, use 249 as device major to avoid crash on Kernel >=4.1

plan44 opened the pull request.

plan44 committed:
Place mailbox in /dev, which is always of type devtmpfs. /tmp might be a file system that does not allow creating character devices.

plan44 committed:
mailbox.c: fixed calculations for non-page-aligned memory mapping requests.

In particular, munmap() often failed because in case of a non-page-aligned mapping, the value passed to munmap() was different from what mmap() originally returned. This caused programs to exit with -1 at ws2811_fini(). This is a problem when exit codes are actually evaluated.

Gadgetoid commented:
I've started getting reports of issues with this library and permissions errors/missing devices in /tmp, is this fix somewhat related? I'm going to give it a try in our Unicorn HAT library either way.

tejonbiker commented:
I have the same behavior, seems like the program can create the /tmp/mailbox-XXXX but can't open it. I added this code to the mbox_open() function in mailbox.c:

if (file_desc < 0) {
      sprintf(system_qry,"file %s",filename);
      system(system_qry);
      printf("Can't open device file %s: %m\n", filename);
      unlink(filename);
      return -1;
   }

And return this:

/tmp/mailbox-3725: character special
Can't open device file /tmp/mailbox-3725: No such device or address.

tejonbiker commented:
Sorry, I forgot mention my kernel info:

uname -a
Linux raspberrypi 4.1.6-v7+ #810 SMP PREEMPT Tue Aug 18 15:32:12 BST 2015 armv7l GNU/Linux

plan44 commented:
@Gadgetoid yes, this fix is related. The regular tmpfs (ram disk file system) cannot have character device nodes, so if /tmp is mapped to a tmpfs (as it is on minibian, a stripped down flavor of the Raspian distribution) the original library does not work.
However, /dev seems always located on a special flavor of tmpfs called devtmpfs which can have character device nodes, so it works for the mailbox.

Gadgetoid commented:
Looks like this fix is related to the Minibian issue, but not to the new one that has developed as of Kernel 4.1.6 which is what @tejonbiker is referring to. I've updated the firmware on a Pi and confirmed the problem- since there's no way of raising issues against this repository, and it's off-topic for this PR, I've raised a new one upstream for discussion: https://github.com/jgarff/rpi_ws281x/issues/41

plan44 committed:
Library crashes on Linux Kernels 4.1 onwards, because character device for mailbox must not use reserved major number 100 (telephony for linux).

Fix is to use a number in the 240..254 range (local/experimental use), as a commenter on github recommended, I chose 249.

See "devices.txt" Linux doc file for a list of device major numbers

ghost commented:
Take a look at [this](https://github.com/joan2937/pigpio/commit/162287708802f379921486865c5c29af3cdba0e8), with the changes from the link rpi_ws281x works fine for me.
